File: src/de/saschahlusiak/freebloks/controller/PlayerData.java
package de.saschahlusiak.freebloks.controller;

import de.saschahlusiak.freebloks.model.Player;

public class PlayerData implements Comparable<PlayerData> {
	public int player1, player2;
	public int place;
	public int points, stones_left;
	public int bonus;
	public boolean is_local, is_perfect;
	
	PlayerData(Spielleiter spiel, int player) {
		this.place = -1;
		this.player1 = player;
		this.player2 = -1;
		this.is_local = spiel.is_local_player(player);
		this.is_perfect = true;
		addPoints(spiel.get_player(player));
	}
	
	PlayerData(Spielleiter spiel, int player1, int player2) {
		this.place = -1;
		this.player1 = player1;
		this.player2 = player2;
		this.is_local = spiel.is_local_player(player1);
		this.is_perfect = true;
		
		addPoints(spiel.get_player(player1));
		addPoints(spiel.get_player(player2));
	}
	
	void addPoints(Player p) {
		this.points += p.m_stone_points;
		this.stones_left += p.m_stone_count;
		if (p.m_stone_count == 0 && p.m_lastStone != null) {
			if (p.m_lastStone.get_stone_shape() == 0) {
				bonus += 20;
			}
			else {
				bonus += 15;
				is_perfect = false;
			}
		} else
			is_perfect = false;
	}

	@Override
	public int compareTo(PlayerData another) {
		if (points > another.points)
			return -1;
		if (points < another.points)
			return 1;
		if (stones_left < another.stones_left)
			return -1;
		if (stones_left > another.stones_left)
			return 1;
		return 0;
	}
}

File: src/main/java/de/saschahlusiak/freebloks/model/PlayerData.kt
package de.saschahlusiak.freebloks.model

class PlayerData @JvmOverloads constructor(spiel: Spielleiter, val player1: Int, val player2: Int = -1) : Comparable<PlayerData> {
    val isLocal: Boolean = spiel.is_local_player(player1)

    // the place will be set afterwards, when all points of all players are known
	var place = -1

    var points: Int
        private set

	var stonesLeft: Int
        private set

	var bonus = 0
        private set

    var isPerfect: Boolean
        private set

    init {
        isPerfect = false

        points = 0
        stonesLeft = 0

        addPoints(spiel.getPlayer(player1))
        // todo: make nullable instead
        if (player2 >= 0)
            addPoints(spiel.getPlayer(player2))
    }

    private fun addPoints(p: Player) {
        points += p.totalPoints
        stonesLeft += p.stonesLeft

        val lastPlayedShape = p.lastShape ?: return
        if (p.stonesLeft == 0) {
            if (lastPlayedShape.size == 1) {
                bonus += 20
                isPerfect = true
            } else {
                bonus += 15
                isPerfect = false
            }
        }
    }

    override fun compareTo(other: PlayerData): Int {
        if (points > other.points) return -1
        if (points < other.points) return 1
        if (stonesLeft < other.stonesLeft) return -1
        return if (stonesLeft > other.stonesLeft) 1 else 0
    }
}